source.go

```diff
@@ -43,18 +56,215 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+	}
+
+	return s.nextObject(ctx)
+}
+
+func (s *Source) nextObject(ctx context.Context) (sdk.Record, error) {
+	if len(s.fetchIDs) == 0 {
+		return sdk.Record{}, errors.New("no page IDs available")
+	}
+	pageID := s.fetchIDs[0]
+	s.fetchIDs = s.fetchIDs[1:]
+
+	// todo support databases
+	sdk.Logger(ctx).Debug().
+		Str("page_id", pageID).
+		Msg("fetching page")
+
+	block, err := s.client.Block.Get(ctx, notion.BlockID(pageID))
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching page %v: %w", pageID, err)
+	}
+
+	children, err := s.getChildren(ctx, block)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching blocks for %v: %w", pageID, err)
+	}
+
+	record, err := s.blockToRecord(block, children)
+	if err != nil {
+		return sdk.Record{}, err
+	}
+	s.lastEditedTime = *block.GetLastEditedTime()
+	return record, nil
+}
+
+// getChildren gets all the child and grand-child blocks of the input block
+func (s *Source) getChildren(ctx context.Context, block notion.Block) ([]notion.Block, error) {
+	var children []notion.Block
+	if !block.GetHasChildren() {
+		sdk.Logger(ctx).Debug().
+			Str("block_id", block.GetID().String()).
+			Msg("block has no children")
+		return children, nil
+	}
+
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		resp, err := s.client.Block.GetChildren(
+			ctx,
+			block.GetID(),
+			&notion.Pagination{
+				StartCursor: cursor,
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed getting children for block ID %v, cursor %v: %w",
+				block.GetID(),
+				cursor,
+				err,
+			)
+		}
+
+		// get grandchildren as well
+		for _, child := range resp.Results {
+			children = append(children, child)
+			grandChildren, err := s.getChildren(ctx, child)
+			if err != nil {
+				return nil, err
+			}
+			children = append(children, grandChildren...)
+		}
+
+		fetch = resp.HasMore
+		cursor = notion.Cursor(resp.NextCursor)
+	}
+	return children, nil
 }
 
-func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
+func (s *Source) Ack(context.Context, sdk.Position) error {
 	return nil
 }
 
-func (s *Source) Teardown(ctx context.Context) error {
+func (s *Source) Teardown(context.Context) error {
 	return nil
 }
+
+func (s *Source) populateIDs(ctx context.Context) error {
+	if len(s.fetchIDs) > 0 {
+		return nil
+	}
+	// the first read attempt (when the connector starts)
+	if !s.firstFetch {
+		sdk.Logger(ctx).Debug().
+			Dur("poll_interval", s.config.pollInterval).
+			Msg("sleeping before checking for changes")
+		time.Sleep(s.config.pollInterval)
+	}
+	s.firstFetch = false
+
+	sdk.Logger(ctx).Debug().Msg("populating IDs")
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		results, err := s.getPages(ctx, cursor)
+		if err != nil {
+			return fmt.Errorf("search failed: %w", err)
+		}
+		s.addToFetchIDs(ctx, results)
+
+		fetch = results.HasMore
+		cursor = results.NextCursor
+	}
+
+	sdk.Logger(ctx).Info().Msgf("fetched %v IDs", len(s.fetchIDs))
```

Comment:
they haven't been fetched at this point yet, right - just added to the fetch IDs slice?

Reply:
actually - I think I'm wrong. reading more closely

Reply:
yeah - I misunderstood. we fetched the record, but soon will convert the block to an SDK record. this makes sense!

Reply:
You're correct in that the pages haven't been fetched at this point, only the IDs. If you need to read more closely, then I need to add more comments.:)